Source/ARTRealtimePresence.m

```diff
@@ -780,85 +778,130 @@ - (void)reenterInternalMembers {
-                messageCopy.action = ARTPresencePresent;
-                // intentional fallthrough
             case ARTPresencePresent:
-                [self internalAdd:messageCopy];
+                messageCopy.action = ARTPresencePresent;
+                [self addInternalMember:messageCopy];
                 break;
             case ARTPresenceLeave:
-                [self internalRemove:messageCopy];
+                if (!message.isSynthesized) {
+                    [self removeInternalMember:messageCopy];
+                }
                 break;
             default:
                 break;
         }
-        return YES;
     }
-    ARTLogDebug(_logger, @"Presence member \"%@\" with action %@ has been ignored", message.memberKey, ARTPresenceActionToStr(message.action));
-    latest.syncSessionId = _syncSessionId;
-    return NO;
-}
+    
+    BOOL memberUpdated = false;
+    switch (message.action) {
+        case ARTPresenceEnter:
+        case ARTPresenceUpdate:
+        case ARTPresencePresent:
+            messageCopy.action = ARTPresencePresent;
+            memberUpdated = [self addMember:messageCopy];
+            break;
+        case ARTPresenceLeave:
+            if (self.syncInProgress) {
+                messageCopy.action = ARTPresenceAbsent; // RTP2f
+                memberUpdated = [self addMember:messageCopy];
+            } else {
+                memberUpdated = [self removeMember:messageCopy];
+            }
+            break;
+        default:
+            break;
+    }
 
-- (void)internalAdd:(ARTPresenceMessage *)message {
-    [self internalAdd:message withSessionId:_syncSessionId];
+    if (memberUpdated) {
+        [self broadcast:message];
+    }
+    else {
+        ARTLogDebug(_logger, @"Presence member \"%@\" with action %@ has been ignored", message.memberKey, ARTPresenceActionToStr(message.action));
+    }
+    [_beforeSyncMembers removeObjectForKey:message.memberKey]; // RTP19
 }
 
-- (void)internalAdd:(ARTPresenceMessage *)message withSessionId:(NSUInteger)sessionId {
-    message.syncSessionId = sessionId;
-    [_members setObject:message forKey:message.memberKey];
-    // Local member
-    if ([message.connectionId isEqualToString:self.connectionId]) {
-        _internalMembers[message.clientId] = message;
-        ARTLogDebug(_logger, @"local member %@ with action %@ has been added", message.memberKey, ARTPresenceActionToStr(message.action).uppercaseString);
+- (BOOL)member:(ARTPresenceMessage *)msg1 isNewerThan:(ARTPresenceMessage *)msg2 {
+    if ([msg1 isSynthesized] || [msg2 isSynthesized]) {
+        return !msg1.timestamp || msg1.timestamp.timeIntervalSince1970 >= msg2.timestamp.timeIntervalSince1970;
+    }
+    
+    NSInteger msg1Serial = [msg1 msgSerialFromId];
+    NSInteger msg1Index = [msg1 indexFromId];
+    NSInteger msg2Serial = [msg2 msgSerialFromId];
+    NSInteger msg2Index = [msg2 indexFromId];
+    
+    if (msg1Serial == msg2Serial) {
+        return msg1Index > msg2Index;
```

Comment:
```suggestion
        return msg1Index >= msg2Index;
```
as per https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/proto_presence_message.go#L101

Reply:
Why here is it opposite then? https://github.com/ably/ably-cocoa/pull/1891/files/1d2f015ff007537fd66febc6b28f79c7e2e37468#r1539081292

Reply:
[RTP5f](https://sdk.ably.com/builds/ably/specification/main/features/#RTP5f) says "_any members present before and after the sync should not emit presence events, all other changes should be reflected in the PresenceMap and should emit presence events on the channel_"

which can be simplified as

"_only_members_that_changed_between_ATTACHED_states_should_result_in_presence_events_" (part of the name of the test _test__023__Presence__Channel_state_change_side_effects_...)

Making it ">=" leads to [this](https://github.com/ably/ably-cocoa/blob/35427f3120e571f5af922f68da3f99d65e668612/Test/Tests/RealtimeClientPresenceTests.swift#L840) array contains all the messages that haven't changed but still broadcasted. wdyt?

Reply:
Also "...[larger being newer](https://sdk.ably.com/builds/ably/specification/main/features/#RTP2b2) in both cases"

Reply:
Okay, I don't think there will be same indexes for messages. So, will approve for now.